src/Network.cpp

```diff
@@ -755,57 +809,60 @@ T relative_difference(T a, T b) {
     if (std::isnan(a) || std::isnan(b)) {
         return std::numeric_limits<T>::max();
     }
-    // Handle sign difference
-    if (((a < 0) != (b < 0)) && (a != 0) && (b != 0)) {
-        return std::numeric_limits<T>::max();
+
+    constexpr auto small_number = 1e-3f;
+    auto fa = std::fabs(a);
+    auto fb = std::fabs(b);
+
+    if (fa > small_number && fb > small_number) {
+        // Handle sign difference
+        if (((a < 0) != (b < 0)) && (a != 0) && (b != 0)) {
+            return std::numeric_limits<T>::max();
+        }
     }
-    a = std::fabs(a);
-    b = std::fabs(b);
 
     // Handle underflow
-    constexpr float small_number = 1e-3;
-    a = std::max(a, small_number);
-    b = std::max(b, small_number);
+    fa = std::max(fa, small_number);
+    fb = std::max(fb, small_number);
 
-    return std::max(fabs((a - b) / a), fabs((a - b) / b));
+    return std::max(fabs((fa - fb) / fa), fabs((fa - fb) / fb));
 }
 
 void compare_net_outputs(std::vector<float>& data,
                          std::vector<float>& ref) {
     // We accept an error up to 5%, but output values
     // smaller than 1/1000th are "rounded up" for the comparison.
-    constexpr float relative_error = 5e-2;
-    int bad_outputs = 0;
+    constexpr float relative_error = 5e-2f;
     for (auto idx = size_t{0}; idx < data.size(); ++idx) {
         auto err = relative_difference(data[idx], ref[idx]);
         if (err > relative_error) {
-            ++bad_outputs;
-            printf("Error in OpenCL calculation at %zu: expected %f got %f "
-                   "(error=%f%%)\n", idx, ref[idx], data[idx], err * 100.0);
+            printf("Error in OpenCL calculation: expected %f got %f "
+                   "(error=%f%%)\n", ref[idx], data[idx], err * 100.0);
             printf("Update your GPU drivers or reduce the amount of games "
                    "played simultaneously.\n");
+            throw std::runtime_error("OpenCL self-check mismatch.");
         }
     }
-    if (bad_outputs > 3) {
```

Comment:
Why do you change to an immediate error here?

Reply:
I basically tried to align with leela-zero as much as possible.

Reply:
Hmm, I'm not encountering the error. What weights are you using?

Reply:
Weights: e49cdc1755712f94abe6802ad44ce5101e18d43d54d73418cd8b2245cf072525 (which are downloaded from the server).
I am using the following recent driver:
```
Platform Name                                   NVIDIA CUDA
  Platform Vendor                                 NVIDIA Corporation
  Platform Version                                OpenCL 1.2 CUDA 9.1.84
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_byte_addressable_store cl_khr_icd cl_khr_gl_sharing cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_copy_opts cl_nv_create_buffer
  Platform Extensions function suffix             NV

  Platform Name                                   NVIDIA CUDA
Number of devices                                 1
  Device Name                                     GeForce GTX 1080
  Device Vendor                                   NVIDIA Corporation
  Device Vendor ID                                0x10de
  Device Version                                  OpenCL 1.2 CUDA
  Driver Version                                  390.25
```

The error only happens once every ~5-10 matches with leela-chess, but never with with leela-zero. I will check if it also happens on the master branch.

Reply:
It also happens on the master branch (after ~20 matches):
```
Loading networks/e49cdc1755712f94abe6802ad44ce5101e18d43d54d73418cd8b2245cf072525
Created dirs data-1
Error in OpenCL calculation at 3086: expected 0.030670 got 0.036442 (error=18.818627%)
Update your GPU drivers or reduce the amount of games played simultaneously.
```

Reply:
It's a [known problem](https://github.com/glinscott/leela-chess/issues/5) for fully random weights that there are big differences between cpu/gpu, presumably because of batchnormalization. But AFAIK no one fully looked into the details.

Reply:
Okay, then we should restore the counter, since the error will happen when starting the distributed setting on random weights.

Reply:
I think restoring the counter is equivalent to reducing the probability of the check occuring no?

Reply:
Indeed, the question is how often this self check would fail, if a wrong driver is used.

Reply:
If we remove the counter, random games will fail as @kiudee noticed.  I think we will need to leave it in until the weights are a little less random :).

Reply:
I think setting `SELFCHECK_PROBABILITY = 6000` is equivalent. See https://github.com/Error323/leela-chess/blob/master/src/Network.cpp#L857

Reply:
I think SELFCHECK_PROBABILTY controls how often compare_net_outputs is called. bad_outputs counter allows up to 3 separate outputs on the same single compare_net_output to differ by more than relative_error (5%). So they're not equivalent.

Reply:
Yeah that's true, OK so should I revert it? I'm not really sure how useful it is I think we should look into the reason why it happens.

Reply:
If we can track the reason down, that would be great.  However, there might just be some problem with the random network weights as noticed in #68.